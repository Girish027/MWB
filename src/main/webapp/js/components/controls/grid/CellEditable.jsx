import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import store from 'state/configureStore';
import * as actionsCellEditable from 'state/actions/actions_cellEditable';
import * as actionsCellEditableManualTagSuggest from 'state/actions/actions_cellEditableManualTagSuggest';

export class CellEditable extends Component {
  constructor(props, context) {
    super(props, context);

    this.onClick = this.onClick.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onChange = this.onChange.bind(this);

    this.isEscaped = false;

    this.state = {
      value: this.props.value,
      isFocused: false,
      isValid: true,
      isTouched: false,
    };
  }

  componentWillReceiveProps(nextProps) {
    const prevValue = this.props.value;
    this.props = nextProps;

    const {
      stateKey, rowIndex, columnIndex, value, tableState,
    } = this.props;
    const { activeRowIndex, activeColumnIndex } = tableState;

    const newState = {};

    if (prevValue != value) {
      newState.value = value;
      this.validate(newState.value);
    }

    if (
      activeRowIndex === rowIndex
            && activeColumnIndex === columnIndex
            && this.props.tableState.value
            && newState.value != this.props.tableState.value
    ) {
      newState.value = this.props.tableState.value;
      this.validate(newState.value);
      store.dispatch(actionsCellEditable.updateValue({
        stateKey,
        value: null,
      }));
    }

    this.checkFocus();
    this.setState(newState);
  }

  componentWillMount() {

  }

  componentDidMount() {
    this.checkFocus();
  }

  onClick(event) {
    const { stopClickPropagation } = this.props;
    if (stopClickPropagation) {
      event.stopPropagation();
    }
  }

  onFocus() {
    this.isEscaped = false;
    this.setState({
      isFocused: true,
      value: this.props.value,
      isValid: true,
    });
    this.dispatchSetActiveCell(true);
  }

  onBlur() {
    let { isValid, value } = this.state;
    if (this.state.value !== this.props.value && !this.isEscaped && this.props.onChange) {
      this.isEscaped = false;
      isValid = this.validate(this.state.value);
      if (isValid) {
        this.props.onChange(value);
      } else {
        value = this.props.value;
      }
    }
    this.setState({
      isFocused: false,
      isTouched: false,
      value,
      isValid,
    });
    this.dispatchSetActiveCell(false);
  }

  onKeyDown(event) {
    const {
      tableState, stateKey, rowIndex, columnIndex, maxRowIndex,
      maxColumnIndex, suggested, activateNextCellOnEnter,
    } = this.props;
    const { activeRowIndex, activeColumnIndex } = tableState;
    const { keyCode, shiftKey, ctrlKey } = event;
    const { input } = this.refs;

    if (event.key === 'Enter') {
      if (activateNextCellOnEnter) {
        let newRowIndex = rowIndex;
        const newColumnIndex = columnIndex;
        newRowIndex++;
        if (newRowIndex > maxRowIndex) {
          newRowIndex = maxRowIndex;
        }
        if (activeRowIndex !== newRowIndex || activeColumnIndex !== columnIndex) {
          if (input) {
            input.blur();
            // this.onBlur();
          }
          store.dispatch(actionsCellEditable.setActiveCell({
            stateKey,
            activeRowIndex: newRowIndex,
            activeColumnIndex: newColumnIndex,
          }));
        }
      } else {
        if (input) {
          input.blur();
          // this.onBlur();
        }
        store.dispatch(actionsCellEditable.setActiveCell({
          stateKey,
          activeRowIndex: null,
          activeColumnIndex: null,
        }));
      }
    } else if (
      keyCode === 9 // Tab key
            || keyCode === 40 // Down Arrow
            || keyCode === 38 // Up Arrow
            || keyCode === 37 // Left Arrow
            || keyCode === 39 // Right Arrow
    ) {
      if (
        suggested
                && (event.keyCode == 40 || event.keyCode == 38)
                && suggested.suggestedTags.length
                && suggested.intent.length
      ) {
        if (event.keyCode == 40) {
          store.dispatch(actionsCellEditableManualTagSuggest.cursorDown({ stateKey }));
        } else if (event.keyCode == 38) {
          store.dispatch(actionsCellEditableManualTagSuggest.cursorUp({ stateKey }));
        }
        return;
      }

      if (typeof rowIndex !== 'undefined' && typeof columnIndex !== 'undefined') {
        let newRowIndex = rowIndex;
        let newColumnIndex = columnIndex;

        const left = (ctrlKey) => {
          if (ctrlKey) {
            newColumnIndex = 0;
          } else {
            newColumnIndex--;
            if (newColumnIndex < 0) {
              newColumnIndex = maxColumnIndex;
              newRowIndex--;
            }
            if (newRowIndex < 0) {
              newRowIndex = 0;
              newColumnIndex = 0;
            }
          }
        };
        const down = (ctrlKey) => {
          if (ctrlKey) {
            newRowIndex = maxRowIndex;
          } else {
            newRowIndex++;
            if (newRowIndex > maxRowIndex) {
              newRowIndex = maxRowIndex;
            }
          }
        };
        const up = (ctrlKey) => {
          if (ctrlKey) {
            newRowIndex = 0;
          } else {
            newRowIndex--;
            if (newRowIndex < 0) {
              newRowIndex = 0;
            }
          }
        };
        const right = (ctrlKey) => {
          if (ctrlKey) {
            newColumnIndex = maxColumnIndex;
          } else {
            newColumnIndex++;
            if (newColumnIndex > maxColumnIndex) {
              newColumnIndex = 0;
              newRowIndex++;
            }
            if (newRowIndex > maxRowIndex) {
              newRowIndex = maxRowIndex;
              newColumnIndex = maxColumnIndex;
            }
          }
        };

        switch (keyCode) {
        case 9: // Tab key
          if (shiftKey) {
            left(false);
          } else {
            right(false);
          }
          break;

        case 40: // Down Arrow
          down(ctrlKey);
          break;

        case 38: // Up Arrow
          up(ctrlKey);
          break;

        case 37: // Left Arrow
          left(ctrlKey);
          break;

        case 39: // Right Arrow
          right(ctrlKey);
          break;
        }

        if (activeRowIndex !== newRowIndex || activeColumnIndex !== newColumnIndex) {
          event.preventDefault();
          if (input) {
            input.blur();
            // this.onBlur();
          }
          store.dispatch(actionsCellEditable.setActiveCell({
            stateKey,
            activeRowIndex: newRowIndex,
            activeColumnIndex: newColumnIndex,
          }));
        }
      }
    } else if (event.keyCode === 27) { // Esc
      this.isEscaped = true;
      if (input) {
        this.setState({
          value: this.props.value,
        });
        input.blur();
        // this.onBlur();
      }
      store.dispatch(actionsCellEditable.setActiveCell({
        stateKey,
        activeRowIndex: null,
        activeColumnIndex: null,
      }));
    }
  }

  onChange(event) {
    const { onValidChange, onEdit } = this.props;
    const { isValid } = this.state;
    const newValue = event.target.value;
    const newIsValid = this.validate(newValue);
    const isValueChanged = newValue !== this.state.value;

    this.setState({
      value: newValue,
      isTouched: true,
      isValid: newIsValid,
    });

    if (isValueChanged && onEdit) {
      onEdit(newValue, newIsValid);
    }

    if (newIsValid != isValid && onValidChange) {
      onValidChange(newIsValid);
    }
  }

  validate(value) {
    const { validation } = this.props;
    if (!validation) {
      return true;
    }
    return validation(value);
  }

  checkFocus() {
    const { input } = this.refs;
    const { tableState, rowIndex, columnIndex } = this.props;
    const { activeRowIndex, activeColumnIndex } = tableState;

    if (!input || typeof rowIndex === 'undefined' || typeof columnIndex === 'undefined') {
      return;
    }

    if (!this.state.isFocused) {
      if (rowIndex === activeRowIndex && columnIndex === activeColumnIndex) {
        input.focus();
        this.onFocus();
      }
    } else if (rowIndex !== activeRowIndex || columnIndex !== activeColumnIndex) {
      input.blur();
      // this.onBlur();
    }
  }

  dispatchSetActiveCell(isActive) {
    const {
      stateKey, tableState, rowIndex, columnIndex,
    } = this.props;
    const { activeRowIndex, activeColumnIndex } = tableState;

    if (typeof rowIndex === 'undefined' || typeof columnIndex === 'undefined') {
      return;
    }

    if (isActive) {
      if (activeRowIndex !== rowIndex || activeColumnIndex !== columnIndex) {
        store.dispatch(actionsCellEditable.setActiveCell({
          stateKey,
          activeRowIndex: rowIndex,
          activeColumnIndex: columnIndex,
        }));
      }
    } else if (activeRowIndex === rowIndex && activeColumnIndex === columnIndex) {
      store.dispatch(actionsCellEditable.setActiveCell({
        stateKey,
        activeRowIndex: null,
        activeColumnIndex: null,
      }));
    }
  }

  render() {
    const { id, className } = this.props;
    const { value, isValid, isTouched } = this.state;

    const classNameAttr = `CellEditable${
      isTouched ? ' touched' : ''
    }${isValid ? '' : ' error'
    }${className ? ` ${className}` : ''}`;

    return (
      <div
        id={id}
        className={classNameAttr}
      >
        <input
          type="text"
          value={value}
          onClick={this.onClick}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onKeyDown={this.onKeyDown}
          onChange={this.onChange}
          ref="input"
        />
      </div>
    );
  }
}

CellEditable.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  stateKey: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onEdit: PropTypes.func,
  rowIndex: PropTypes.number,
  columnIndex: PropTypes.number,
  maxRowIndex: PropTypes.number,
  maxColumnIndex: PropTypes.number,
  stopClickPropagation: PropTypes.bool,
  validation: PropTypes.func,
  onValidChange: PropTypes.func,
  suggestedStateKey: PropTypes.string,
  activateNextCellOnEnter: PropTypes.bool,
};

CellEditable.defaultProps = {
  value: '',
  rowIndex: -1,
  columnIndex: -1,
  activateNextCellOnEnter: true,
};

const mapStateToProps = (state, ownProps) => ({
  tableState: state.cellEditable ? state.cellEditable.get(ownProps.stateKey) : {},
  suggested: ownProps.suggestedStateKey && typeof state[ownProps.suggestedStateKey] !== 'undefined' ? state[ownProps.suggestedStateKey].get(ownProps.stateKey) : null,
});

export default connect(mapStateToProps)(CellEditable);
