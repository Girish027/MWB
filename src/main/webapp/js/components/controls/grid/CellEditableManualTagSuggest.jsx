import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import store from 'state/configureStore';
import validationUtil from 'utils/ValidationUtil';
import * as actionsCellEditableManualTagSuggest from 'state/actions/actions_cellEditableManualTagSuggest';
import * as actionsCellEditable from 'state/actions/actions_cellEditable';

export class CellEditableManualTagSuggest extends Component {
  constructor(props, context) {
    super(props, context);

    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.handleSuggestHover = this.handleSuggestHover.bind(this);

    this.isValid = true;

    this.state = {
      keyboardSelectedIndex: -1,
      cursor: this.props.suggestions.cursor,
    };
  }

  componentWillReceiveProps(nextProps) {
    const prevIntent = this.props.intent;
    this.props = nextProps;

    const { intent, stateKey, projectId } = this.props;
    this.isValid = validationUtil.validatePartialTaggingGuideTag(intent);

    if (this.isValid && intent.length > 0 && prevIntent != intent) {
      store.dispatch(actionsCellEditableManualTagSuggest.getSuggestedTags({
        stateKey, projectId, intent,
      }));
    }

    if (this.props.suggestions.cursor != this.state.cursor) {
      this.moveSelection();
    }
  }

  moveSelection = () => {
    const { stateKey } = this.props;
    const { cursor: updatedCursor, suggestedTags } = this.props.suggestions;
    const { keyboardSelectedIndex, cursor: previousCursor } = this.state;

    if (!suggestedTags.length) {
      return;
    }

    let newIndex = keyboardSelectedIndex;
    if (updatedCursor > previousCursor) {
      newIndex += 1;
      if (newIndex >= suggestedTags.length) {
        newIndex = 0; // first of the list
      }
    } else {
      newIndex -= 1;
      if (newIndex === -1) {
        newIndex = suggestedTags.length - 1; // last of the list
      }
    }


    const container = document.getElementsByClassName('SuggestContainer')[0];
    const currentRow = document.getElementsByClassName('Suggest')[newIndex];
    if (currentRow) {
      const tag = currentRow.innerText;

      store.dispatch(actionsCellEditable.updateValue({
        stateKey,
        value: tag,
      }));

      container.scrollTop = currentRow.offsetTop;
    }

    this.setState({
      cursor: updatedCursor,
      keyboardSelectedIndex: newIndex,
    });
  }

  handleSuggestHover(event) {
    const { stateKey } = this.props;
    const tag = event.target.getAttribute('data-label');

    store.dispatch(actionsCellEditable.updateValue({
      stateKey,
      value: tag,
    }));
  }

  handleKeyPress(event) {
    const { suggestedTags } = this.props.suggestions;
    if (!suggestedTags.length) {
      return;
    }
    let { keyboardSelectedIndex } = this.state;

    if (event.key === 'Up') {
      keyboardSelectedIndex--;
      if (keyboardSelectedIndex < 0) {
        keyboardSelectedIndex = suggestedTags.length - 1;
      }
    } else if (event.key === 'Down') {
      keyboardSelectedIndex++;
      if (keyboardSelectedIndex > suggestedTags.length - 1) {
        keyboardSelectedIndex = 0;
      }
    }

    this.setState({
      keyboardSelectedIndex,
    });
  }

  render() {
    const { id, className, highlightTags } = this.props;
    const { intent, suggestedTags } = this.props.suggestions;
    const { keyboardSelectedIndex } = this.state;

    if (this.isValid && intent != this.props.intent) {
      return null;
    }

    let listItems = [];
    if (!this.isValid) {
      listItems = [
        <div
          key="cell-0"
          className="Suggest error"
        >
                    No matching tags
        </div>,
        <div
          key="cell-1"
          className="Suggest error"
        >
                    To create a new tag, please follow the topic-goal format
        </div>,
      ];
    } else if (intent.length && !suggestedTags.length) {
      listItems = [
        <div
          key="cell-0"
          className="Suggest warning"
        >
                    No matching tags
        </div>,
        <div
          key="cell-1"
          className="Suggest warning"
        >
                    This will create a new tag
        </div>,
      ];
    } else if (suggestedTags.length) {
      const tagsOrderred = [];
      suggestedTags.forEach((tag) => {
        if (highlightTags.indexOf(tag) != -1) {
          tagsOrderred.unshift({ tag, isSuggested: true });
        } else {
          tagsOrderred.push({ tag, isSuggested: false });
        }
      });

      tagsOrderred.forEach((data, index) => {
        const itemClassName = `Suggest${
          keyboardSelectedIndex == index ? ' keyboardSelected' : ''
        }${data.isSuggested ? ' suggested' : ''}`;
        listItems.push(<div
          key={`cell-${index}`}
          tabIndex="-1"
          className={itemClassName}
          data-label={data.tag}
          onMouseOver={this.handleSuggestHover}
        >
          {data.tag}
        </div>);
      });
    }

    const classNameAttr = `SuggestContainer${className ? ` ${className}` : ''}`;
    return (
      <div
        id={id}
        className={classNameAttr}
        onKeyPress={this.handleKeyPress}
      >
        { listItems }
      </div>
    );
  }
}

CellEditableManualTagSuggest.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  stateKey: PropTypes.string.isRequired,
  intent: PropTypes.string.isRequired,
  projectId: PropTypes.string.isRequired,
  highlightTags: PropTypes.array,
  rowIndex: PropTypes.number,
  columnIndex: PropTypes.number,
};

CellEditableManualTagSuggest.defaultProps = {
  highlightTags: [],
  rowIndex: -1,
  columnIndex: -1,
};

const mapStateToProps = (state, ownProps) => ({
  suggestions: state.cellEditableManualTagSuggest.get(ownProps.stateKey),
});

export default connect(mapStateToProps)(CellEditableManualTagSuggest);
