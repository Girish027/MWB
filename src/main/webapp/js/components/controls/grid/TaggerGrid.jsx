import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid } from 'react-redux-grid';

export default class TaggerGrid extends Component {
  constructor(props) {
    super(props);
    this.props = props;
  }

  render() {
    const {
      id, className, config, style,
    } = this.props;

    return (
      <div
        id={id}
        className={`TaggerGrid${className ? ` ${className}` : ''}`}
        style={style}
      >
        <Grid {...config} data={this.props.data} />
      </div>
    );
  }
}

TaggerGrid.propTypes = {
  config: PropTypes.object.isRequired,
  id: PropTypes.string,
  className: PropTypes.string,
  data: PropTypes.array,
  style: PropTypes.object,
};

TaggerGrid.defaultProps = {

};
