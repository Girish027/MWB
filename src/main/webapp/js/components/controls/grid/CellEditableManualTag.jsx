import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import validationUtil from 'utils/ValidationUtil';
import CellEditable from 'components/controls/grid/CellEditable';
import CellEditableManualTagSuggest from 'components/controls/grid/CellEditableManualTagSuggest';

export class CellEditableManualTag extends Component {
  constructor(props, context) {
    super(props, context);
    this.onEdit = this.onEdit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onValidChange = this.onValidChange.bind(this);

    this.state = {
      editedValue: this.props.value,
      isEditedValueValid: true,
    };
  }

  componentWillReceiveProps(nextProps) {
    const prevValue = this.props.value;
    this.props = nextProps;

    if (prevValue !== this.props.value) {
      this.setState({
        editedValue: this.props.value,
        isEditedValueValid: true,
      });
    }
  }

  onEdit(newValue, isValid) {
    this.setState({
      editedValue: newValue,
    });
  }

  onChange(newValue) {
    const { onChange } = this.props;
    if (onChange) {
      onChange(newValue);
    }
  }

  onValidChange(isValid) {
  }

  render() {
    const {
      id, className, stateKey, value, rowIndex, columnIndex,
      maxRowIndex, maxColumnIndex, stopClickPropagation,
      projectId, editable, highlightTags, activateNextCellOnEnter,
    } = this.props;
    const { editedValue } = this.state;

    const isActive = typeof rowIndex !== 'undefined'
            && typeof columnIndex !== 'undefined'
            && editable.activeRowIndex === rowIndex
            && editable.activeColumnIndex === columnIndex;

    return (
      <div
        id={id}
        className={`CellEditableManualTag${className ? ` ${className}` : ''}`}
      >
        <CellEditable
          stateKey={stateKey}
          value={value}
          rowIndex={rowIndex}
          columnIndex={columnIndex}
          maxRowIndex={maxRowIndex}
          maxColumnIndex={maxColumnIndex}
          stopClickPropagation={stopClickPropagation}
          validation={validationUtil.validateTaggingGuideTag}
          suggestedStateKey="cellEditableManualTagSuggest"
          activateNextCellOnEnter={activateNextCellOnEnter}
          onValidChange={this.onValidChange}
          onChange={this.onChange}
          onEdit={this.onEdit}
        />
        {isActive && editedValue.length ? (
          <CellEditableManualTagSuggest
            stateKey={stateKey}
            intent={editedValue}
            projectId={projectId}
            highlightTags={highlightTags}
          />
        ) : null}
      </div>
    );
  }
}

CellEditableManualTag.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  stateKey: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  projectId: PropTypes.string.isRequired,
  rowIndex: PropTypes.number,
  columnIndex: PropTypes.number,
  maxRowIndex: PropTypes.number,
  maxColumnIndex: PropTypes.number,
  stopClickPropagation: PropTypes.bool,
  highlightTags: PropTypes.array,
  onChange: PropTypes.func,
  activateNextCellOnEnter: PropTypes.bool,
};

CellEditableManualTag.defaultProps = {
  value: '',
  rowIndex: -1,
  columnIndex: -1,
  highlightTags: [],
  activateNextCellOnEnter: true,
};

const mapStateToProps = (state, ownProps) => ({
  editable: state.cellEditable.get(ownProps.stateKey),
  suggest: state.cellEditableManualTagSuggest.get(ownProps.stateKey),
});

export default connect(mapStateToProps)(CellEditableManualTag);
