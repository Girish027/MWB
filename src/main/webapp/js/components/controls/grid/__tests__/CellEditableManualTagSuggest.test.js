import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { CellEditableManualTagSuggest } from 'components/controls/grid/CellEditableManualTagSuggest';

describe('<CellEditableManualTagSuggest />', () => {
  let wrapper;

  const props = {
    dispatch: jest.fn(),
    stateKey: 'sdffds',
    projectId: '23',
    intent: 'sdff',
    suggestions: {
      cursor: 2,
      intent: 'sdff',
      suggestedTags: [],
    },
  };

  const getShallowWrapper = (propsObj) => shallow(<CellEditableManualTagSuggest
    {...propsObj}
  />);

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = mount(<CellEditableManualTagSuggest
        {...props}
      />);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly for default props', () => {
      const defaultProps = {
        ...props,
      };
      wrapper = getShallowWrapper(defaultProps);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly for default props with suggestedTags not empty', () => {
      const defaultProps = {
        ...props,
        suggestions: {
          intent: 'sdff',
          suggestedTags: ['ddf'],
        },
      };
      wrapper = getShallowWrapper(defaultProps);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly for default props with highlightTags', () => {
      const defaultProps = {
        ...props,
        highlightTags: ['ddf'],
        suggestions: {
          intent: 'sdff',
          suggestedTags: ['ddf'],
        },
      };
      wrapper = getShallowWrapper(defaultProps);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    describe('handleSuggestHover', () => {
      test('should call handleSuggestHover with event', () => {
        const event = {
          target: {
            getAttribute: jest.fn().mockImplementation(() => ('tags')),
          },
        };
        wrapper = getShallowWrapper(props);
        wrapper.instance().handleSuggestHover(event);
      });
    });

    describe('handleKeyPress', () => {
      test('should call handleKeyPress with no event', () => {
        const event = {
        };
        wrapper = getShallowWrapper(props);
        wrapper.instance().handleKeyPress(event);
      });

      test('should call handleKeyPress with event key as Up', () => {
        const defaultProps = {
          ...props,
          highlightTags: ['ddf'],
          suggestions: {
            cursor: 1,
            intent: 'sdff',
            suggestedTags: ['ddf'],
          },
        };
        const event = {
          key: 'Up',
        };
        wrapper = getShallowWrapper(defaultProps);
        wrapper.setState({ keyboardSelectedIndex: 0 });
        wrapper.instance().handleKeyPress(event);
        expect(wrapper.state().keyboardSelectedIndex).toEqual(0);
      });

      test('should call handleKeyPress with event key as Down case 1', () => {
        const defaultProps = {
          ...props,
          highlightTags: ['ddf'],
          suggestions: {
            cursor: 1,
            intent: 'sdff',
            suggestedTags: ['ddf'],
          },
        };
        const event = {
          key: 'Down',
        };
        wrapper = getShallowWrapper(defaultProps);
        wrapper.setState({ keyboardSelectedIndex: 1 });
        wrapper.instance().handleKeyPress(event);
        expect(wrapper.state().keyboardSelectedIndex).toEqual(0);
      });

      test('should call handleKeyPress with event key as Down case 2', () => {
        const defaultProps = {
          ...props,
          highlightTags: ['ddf'],
          suggestions: {
            cursor: 1,
            intent: 'sdff',
            suggestedTags: ['ddf', 'sfdsfdf'],
          },
        };
        const event = {
          key: 'Down',
        };
        wrapper = getShallowWrapper(defaultProps);
        wrapper.setState({ keyboardSelectedIndex: 0 });
        wrapper.instance().handleKeyPress(event);
        expect(wrapper.state().keyboardSelectedIndex).toEqual(1);
      });
    });

    describe('moveSelection', () => {
      test('should call moveSelection', () => {
        wrapper = getShallowWrapper(props);
        wrapper.setState({ keyboardSelectedIndex: 1, cursor: 1 });
        wrapper.instance().moveSelection();
      });

      test('should call moveSelection with suggestions cursor is same as state cursor', () => {
        const defaultProps = {
          ...props,
          highlightTags: ['ddf'],
          suggestions: {
            cursor: 1,
            intent: 'sdff',
            suggestedTags: ['ddf'],
          },
        };
        wrapper = getShallowWrapper(defaultProps);
        wrapper.setState({ keyboardSelectedIndex: 0, cursor: 1 });
        wrapper.instance().moveSelection();
        expect(wrapper.state().cursor).toEqual(1);
      });

      test('should call moveSelection with suggestedTags', () => {
        const defaultProps = {
          ...props,
          highlightTags: ['ddf'],
          suggestions: {
            cursor: 2,
            intent: 'sdff',
            suggestedTags: ['ddf'],
          },
        };
        wrapper = getShallowWrapper(defaultProps);
        wrapper.setState({ keyboardSelectedIndex: 1, cursor: 1 });
        wrapper.instance().moveSelection();
        expect(wrapper.state().cursor).toEqual(2);
      });
    });

    describe('componentWillReceiveProps', () => {
      test('should call componentWillReceiveProps', () => {
        const nextProps = {
          intent: 'ffds',
          stateKey: 'dsf',
          projectId: '23',
          suggestions: {
            cursor: 2,
          },
        };
        wrapper = getShallowWrapper(props);
        wrapper.setState({ keyboardSelectedIndex: 1, cursor: 1 });
        wrapper.instance().moveSelection = jest.fn();
        wrapper.instance().componentWillReceiveProps(nextProps);
        expect(wrapper.instance().moveSelection).toHaveBeenCalledWith();
      });
    });
  });
});
