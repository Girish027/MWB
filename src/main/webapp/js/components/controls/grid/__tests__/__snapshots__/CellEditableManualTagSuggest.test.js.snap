// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CellEditableManualTagSuggest /> Snapshots renders correctly for default props 1`] = `
<div
  className="SuggestContainer"
  onKeyPress={[Function]}
>
  <div
    className="Suggest warning"
    key="cell-0"
  >
    No matching tags
  </div>
  <div
    className="Suggest warning"
    key="cell-1"
  >
    This will create a new tag
  </div>
</div>
`;

exports[`<CellEditableManualTagSuggest /> Snapshots renders correctly for default props with highlightTags 1`] = `
<div
  className="SuggestContainer"
  onKeyPress={[Function]}
>
  <div
    className="Suggest suggested"
    data-label="ddf"
    key="cell-0"
    onMouseOver={[Function]}
    tabIndex="-1"
  >
    ddf
  </div>
</div>
`;

exports[`<CellEditableManualTagSuggest /> Snapshots renders correctly for default props with suggestedTags not empty 1`] = `
<div
  className="SuggestContainer"
  onKeyPress={[Function]}
>
  <div
    className="Suggest"
    data-label="ddf"
    key="cell-0"
    onMouseOver={[Function]}
    tabIndex="-1"
  >
    ddf
  </div>
</div>
`;
